### PR TITLE
03 02 methods and oop

### PR DESCRIPTION
**Pull request by igorchupin**
1) Now Product Variable inherits the Product and contains fields Rate and Name.

2) Now Methods from "StoreApp" (print store) were moved to the corresponding classes (ProductVariable, Store)

### CODE REVIEW

File: store/src/main/java/Product.java [hunk 1]
```diff
@@ -3,17 +3,22 @@
 
 public class Product {
     ArrayList<ProductVariable> productVariablesList;
-    private String name;
-    private Integer rate;
+    String name;
```
Comment (shukal94) [inline store/src/main/java/Product.java:6]: why did you remove private fields?

File: store/src/main/java/Product.java [hunk 2]
```diff
@@ -3,17 +3,22 @@
...
 
     public Product(String name, int rate, ArrayList<ProductVariable> productVariables) {
         this.name = name;
         this.rate = rate;
         this.productVariablesList = productVariables;
     }
 
+    public Product(String name, int rate) {
+        this.name = name;
+        this.rate = rate;
```
Comment (shukal94) [inline store/src/main/java/Product.java:17]: if you want to have an additional constructor let's init productVariables as an empty list

File: store/src/main/java/Product.java [hunk 3]
```diff
@@ -3,17 +3,22 @@
...
+    }
+
     public String getName() {
-        return name;
+        return this.name;
```
Comment (shukal94) [inline store/src/main/java/Product.java:21]: you don't need this here

File: store/src/main/java/Product.java [hunk 4]
```diff
@@ -28,7 +33,43 @@ public void setRate(int rate) {
...
+        int result = 1;
+        result = prime * result + name.hashCode();
+        result = prime * result + rate;
+        result = prime * result + productVariablesList.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "Product name: " + this.getName() + "; product rate: " + this.getRate() + "; contains " + this.productVariablesList.size() + " variables";
```
Comment (shukal94) [inline store/src/main/java/Product.java:73]: see comments above about Category class

File: store/src/main/java/ProductVariable.java [hunk 5]
```diff
@@ -1,34 +1,162 @@
...
+import java.util.Comparator;
 
 enum CoversTypes {
     Hard,
     Soft,
     Exclusive
 }
 
-public class ProductVariable {
+public class ProductVariable extends Product implements Comparable<ProductVariable> {
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:11]: why do we need to extend Product?

File: store/src/main/java/ProductVariable.java [hunk 6]
```diff
@@ -1,34 +1,162 @@
...
     private String publisher;
-    private final int quantity;
     private final BigDecimal price;
+    private int quantity = 0;
     CoversTypes coverType;
+    static ArrayList<ProductVariable> listAllVariables = new ArrayList<>();
 
-    public ProductVariable(String publisher, CoversTypes types, BigDecimal price, int left) {
+    public ProductVariable(String name, Integer rate, String publisher, CoversTypes types, BigDecimal price, int left) {
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:18]: different variations may have different rate, or all the variations have one rate
in the second case you don't need to inherit from Product
in the first case you can iniе rate field here

File: store/src/main/java/ProductVariable.java [hunk 7]
```diff
@@ -1,34 +1,162 @@
...
 }
 
-public class ProductVariable {
+public class ProductVariable extends Product implements Comparable<ProductVariable> {
     private String publisher;
-    private final int quantity;
     private final BigDecimal price;
+    private int quantity = 0;
     CoversTypes coverType;
+    static ArrayList<ProductVariable> listAllVariables = new ArrayList<>();
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:16]: why do you need variable list in variable class?

File: store/src/main/java/ProductVariable.java [hunk 8]
```diff
@@ -1,34 +1,162 @@
...
     Soft,
     Exclusive
 }
 
-public class ProductVariable {
+public class ProductVariable extends Product implements Comparable<ProductVariable> {
     private String publisher;
-    private final int quantity;
     private final BigDecimal price;
+    private int quantity = 0;
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:14]: remove init

File: store/src/main/java/ProductVariable.java [hunk 9]
```diff
@@ -1,34 +1,162 @@
...
+    @Override
     public String toString() {
         return " Publisher: " + this.publisher + "; Cover: " + coverType.name() + "; Price: " + this.price +
-                "$; Exemplars left: " + quantity + ";";
+                "$; Exemplars left: " + quantity + ";" + " // Name: " + this.name + "; Rate: " + this.rate + ";";
+    }
+
+    @Override
+    public int compareTo(ProductVariable variableProduct) {
+        if (this.price.doubleValue() == variableProduct.price.doubleValue()) {
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:79]: why don't you use equals from BigDecimal.class?

File: store/src/main/java/ProductVariable.java [hunk 10]
```diff
@@ -1,34 +1,162 @@
...
+                }
+            }
+        }
+        return listAllVariables;
+    }
+
+    public static void printAllVariablesList(ArrayList<ProductVariable> variableList) {
+        for (int i = 0; i < variableList.size(); i++) {
+            System.out.println(i + 1 + ") " + variableList.get(i).toString());
+        }
```
Comment (shukal94) [inline store/src/main/java/ProductVariable.java:160]: please follow single-responsibility principle: remove those methods

File: store/src/main/java/Store.java [hunk 11]
```diff
@@ -21,4 +22,58 @@ public void addCategory(String categoryName) {
...
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + categoryList.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "This shop contains: " + this.categoryList.size() + " categories";
+    }
```
Comment (shukal94) [inline store/src/main/java/Store.java:49]: I think there's no need to override those methods here

File: store/src/main/java/SubCategory.java [hunk 12]
```diff
@@ -27,4 +27,31 @@ public void setName(String name) {
...
+        int result = 1;
+        result = prime * result + name.hashCode();
+        result = prime * result + productList.hashCode();
+        return result;
+    }
+
+
+    @Override
+    public String toString() {
+        return "Subcategory name: " + this.getName() + "; Contains " + this.getProductList().size() + " products";
```
Comment (shukal94) [inline store/src/main/java/SubCategory.java:55]: see the comments for Product

**Review by shukal94 — CHANGES_REQUESTED** [review]

File: store/src/main/java/Category.java [hunk 13]
```diff
@@ -1,7 +1,8 @@
 import java.util.ArrayList;
+import java.util.Objects;
 
 public class Category {
-    private String name;
+    private String name = "";
```
[the remark this answers is not in the document]
Comment (igorchupin) [reply store/src/main/java/Category.java:5]: removed

File: store/src/main/java/Category.java [hunk 14]
```diff
@@ -30,4 +31,30 @@ public void setSubcategoryList(ArrayList<SubCategory> subcategoryList) {
...
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + name.hashCode();
+        result = prime * result + subcategoryList.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "Category name: " + this.getName() + "; Contains " + this.getSubcategoryList().size() + " subcategories";
```
[the remark this answers is not in the document]
Comment (igorchupin) [reply store/src/main/java/Category.java:58]: This was changed

File: store/src/main/java/Product.java (hunk 1 above)
Comment (igorchupin) [reply store/src/main/java/Product.java:6]: Changed

File: store/src/main/java/Product.java (hunk 2 above)
Comment (igorchupin) [reply store/src/main/java/Product.java:17]: Constructor was removed

File: store/src/main/java/Product.java (hunk 3 above)
Comment (igorchupin) [reply store/src/main/java/Product.java:21]: This was changed

File: store/src/main/java/Product.java (hunk 4 above)
Comment (igorchupin) [reply store/src/main/java/Product.java:73]: Changed "toString"

File: store/src/main/java/ProductVariable.java (hunk 5 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:11]: This was removed

File: store/src/main/java/ProductVariable.java (hunk 6 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:18]: all the variations have one rate. Fields "rate" and "name" were added.

File: store/src/main/java/ProductVariable.java (hunk 7 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:16]: We do not need it. I removed it to the "Utils" class

File: store/src/main/java/ProductVariable.java (hunk 8 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:14]: removed

File: store/src/main/java/ProductVariable.java (hunk 9 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:79]: Changed to (this.price.equals(variableProduct.price))

File: store/src/main/java/ProductVariable.java (hunk 10 above)
Comment (igorchupin) [reply store/src/main/java/ProductVariable.java:160]: Moved to the Utils class

File: store/src/main/java/Store.java (hunk 11 above)
Comment (igorchupin) [reply store/src/main/java/Store.java:49]: they were removed

**Review by shukal94 — APPROVED** [review]

**Pull request merged**